Frameworks/Foundation/NSKeyValueCoding.mm

```diff
@@ -578,30 +588,73 @@ - (NSMutableOrderedSet*)mutableOrderedSetValueForKeyPath:(NSString*)keyPath {
     return StubReturn();
 }
 
-/**
- @Status Stub
- @Notes
-*/
-- (void)setNilValueForKey:(NSString*)key {
-    UNIMPLEMENTED();
+static SEL KVCValidatorForPropertyName(NSObject* self, const char* key) {
+    size_t len = strlen(key);
+    // For the key "example", we must construct the following buffer:
+    // _ _ _ _ _ _ _ _ _ x a m p l e _ _ _ _ _ _ _ \0
+    // and fill it with the following characters:
+    // v a l i d a t e E x a m p l e : e r r o r : \0
+    std::unique_ptr<char[]> chars(new char[8 + len + 8]{});
+    strcpy_s(chars.get() + 9, len, key + 1);
+    chars[0] = 'v';
+    chars[1] = 'a';
+    chars[2] = 'l';
+    chars[3] = 'i';
+    chars[4] = 'd';
+    chars[5] = 'a';
+    chars[6] = 't';
+    chars[7] = 'e';
+    chars[8] = toupper(key[0]);
+    chars[8 + len] = ':';
+    chars[8 + len + 1] = 'e';
+    chars[8 + len + 2] = 'r';
+    chars[8 + len + 3] = 'r';
+    chars[8 + len + 4] = 'o';
+    chars[8 + len + 5] = 'r';
+    chars[8 + len + 6] = ':';
+    SEL sel = sel_getUid(chars.get());
+    return ([self respondsToSelector:sel]) ? sel : nullptr;
 }
 
 /**
- @Status Stub
- @Notes
+ @Status Caveat
+ @Notes Invoked validate method must return BOOL
 */
 - (BOOL)validateValue:(id _Nullable*)ioValue forKey:(NSString*)key error:(NSError* _Nullable*)outError {
-    UNIMPLEMENTED();
-    return StubReturn();
+    SEL validationSelector = KVCValidatorForPropertyName(self, [key UTF8String]);
+    if (validationSelector) {
+        Method sig = class_getInstanceMethod([self class], validationSelector);
```

Comment:
we may want to prefer `[class methodSignatureForSelector:]`; this is a safe override point for subclassers.

Some of our other KVC code may run afoul of this.